feat 新規タスク追加機能の追加

YukiYuigishi opened the pull request:
新規タスクを追加する機能です。
- 新規タスクの追加機能
- config.json読み取り機能

kurume-robo committed:
update

kurume-robo committed:
add crate serde

kurume-robo committed:
add type

YukiYuigishi committed:
feat func add_task

YukiYuigishi committed:
feat func add_task

コメントの追加

YukiYuigishi committed:
feat add_task test

動かない

YukiYuigishi committed:
fix add_task() test

mongodbのポート番号にtypoがあったので修正

kurume-robo committed:
Merge remote-tracking branch 'origin/feat/backend' into feat/backend

kurume-robo committed:
feat seard_json crate

kurume-robo committed:
fix Cargo.lock

kurume-robo committed:
wakaran

kurume-robo committed:
Merge remote-tracking branch 'origin/feat/backend' into feat/backend

YukiYuigishi committed:
feat config まだ動かず

YukiYuigishi committed:
WIP

YukiYuigishi committed:
feat read_json_config

YukiYuigishi committed:
Merge remote-tracking branch 'origin/feat/backend' into feat/backend

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+        let database = client.database("taskdb");
+        let collection = database.collection::<Task>("task");
+
+        let docs = vec![new_task];
+        collection.insert_many(docs, None)?;
```
1つしか挿入しないなら [`Collection::insert_one`](https://docs.rs/mongodb/latest/mongodb/struct.Collection.html#method.insert_one) を使えば良いのでは？

YukiYuigishi replied:
１つしか挿入しない想定なので`Collection::insert_one`を使うように変更します。

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+    }
+}
+
+pub mod db_controller {
```
`db_controller` って深い意味はない？一瞬Clean ArchitectureのControllerのことかと思ったけど(深読みしすぎ)

YukiYuigishi replied:
`db_controller`には深い意味はないので、名前が大仰すぎるというなら、変えたほうがいいかもしれないかなとは思ってる。ただ、add_task()以外にも、dlete_task()とか、finish_task()とかを実装しようかなと思ってる。

nanai10a replied:
まぁ普通に `db` でも構わない感あるなぁ 仰々しく言っても `db_repository` とは言わないし… `repository::mongogb` `repo::mongo` やな 仰々しくすれば

YukiYuigishi replied:
`db`でも構わないのか(まともに触ったことなかったから知らなかった)
`db_controller`から`db`に変更します。

nanai10a replied:
> まともに触ったことなかったから

いや, そんな確固たる根拠が有るわけじゃないですが… (rsで書かれたbotとか読んだことあるのapproversのくらいなので)

YukiYuigishi replied:
正確に言うと、バックエンドを書いたことが無かったからっていうのが正しいですね.......

nanai10a replied:
草

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    //新規タスクの追加
+    pub fn add_task(new_task: Task, db_address: String) -> Result<(), mongodb::error::Error> {
```
`db_address` を引数として渡すとtaskを挿入するたびにconnectionを確立しないといけないので *きっと* コストが高いです,
dbへの接続とtaskの挿入は分離したら良いのでは？: `db_controller::add_task(Task, Collection) -> Result<...>` な関数にして, `db_controller::create_connection_with_config(Config) -> Result<Client, ...>` みたいな(**例です**)
`db_controller::add_task` は呼び出し側からするとdbへのconnectionを作成する必要が有るので一手間増えますが, dbへの操作を増やしたとき(例えば `edit_task` , `delete_task` のような)にconnectionを使いまわしできるので, 良いのではないかなぁと

YukiYuigishi replied:
connectionが複数生えるのはよろしくないので、sync::Clientを引数として取るように変更しました。

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+
+    #[test]
+    fn add_new_task() {
+        let new_task = Task {
+            scheduled_date: Utc::now(),
+            register_date: Utc::now(),
+            title: String::from("LLP"),
+            memo: String::from("Life Love Peace"),
+            finish: false,
+        };
```
もしも `Task` を作るときに `scheduled_date` は *いつも* `Utc::new()` , `finish` は *いつも* `false` なら, [`Default`](https://doc.rust-lang.org/stable/core/default/trait.Default.html) の実装をお薦めしたい, 或いは `Task::new(title: String, memo: String) -> Task` を追加するでも良いと思う

前者なら
```rs
let new_task = Task {
    title: String::from("LLP"),
    memo: String::From("Live Love Peace"),
    ..Default::default(),
};
```

後者なら
```rs
let new_task = Task::new(String::from("LLP"), String::from("Life Love Peace"));
```

になります

YukiYuigishi replied:
Defaultを実装しました。

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+        let db_address = String::from("mongodb://localhost:27017");
+
+        add_task(new_task, db_address).expect("Failed to add new Task");
+        println!("add new Task");
```
出力されようがされまいがこのtestは *panicしなければ* 常に正常終了の関数なので `println!` は不要かと

YukiYuigishi replied:
testの場合は標準入出力は行われず、正常終了か否かのみ見ることができるってことでいいの？

nanai10a replied:
標準入出力を表示する方法はあります(異常終了時と, optionを付けた時)が, 別にtestならprintln要る？という感じ
`dbg!` でデータを確認するとかなら分かるけど, これreturn直前の "確実にこの後testは成功します" ってタイミングのマーカー("処理がここを通りましたよ")なので, 必要性を感じないという話です

YukiYuigishi replied:
Defaultを実装しました。

YukiYuigishi replied:
`println!`は消しました。
`add_task()`のエラー処理はexpectでやってしまって大丈夫ですか？

YukiYuigishi replied:
修正
9b4701ac7c22b54417b51723c5f053d737fa5431

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+pub mod config {
```
(関係ないところにrefしてごめんなさい)
`String::from` 大好きなら止めないんですけど, `&str` から `String` への変換は色々あって, 例えば
```rs
let string: String = "using Into::into".into();
```
や
```rs
let string: String = "using ToString::to_string".to_string();
```
があります, 参考までに: [`alloc::string`](https://doc.rust-lang.org/stable/alloc/string/index.html)

ただ一つ言えるのは, *変にごちゃごちゃ* しなければ `&str` -> `String` の変換はちゃんと最小限のheap-allocationで済みます, *ごちゃごちゃ* すると**3回**とか余計なallocationが生み出される(※確保されてすぐ開放される)のでご注意を

YukiYuigishi replied:
&strからの変換方法をあまり知らなかったので助かります。

nanai10a reviewed (changes requested):
Good work 🍵

included-若干-お気持ち
rejectしても構わないですが, 取り敢えず言えるところは言ったつもりです

kawaemon commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+    pub struct Config {
+        pub db_address: String,
+    }
+
+    pub fn read_json_config<P: AsRef<Path>>(path: P) -> Result<Config, Box<dyn Error>> {
+        let file = File::open(path).expect("Cannot read file");
```
で良いと思う
```suggestion
    pub fn read_json_config(path: impl AsRef<Path>) -> Result<Config, Box<dyn Error>> {
        let file = File::open(path).expect("Cannot read file");
```

nanai10a replied:
`impl` 教の人じゃん(※**私も**) generics使っても良いのでは
edit: 書き方の思想の問題感あるので

YukiYuigishi replied:
`impl`で書いた場合、見た目が良くなる(読みやすくなる)以外の利点って何かありますか？

kawaemon commented on src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    pub fn read_json_config<P: AsRef<Path>>(path: P) -> Result<Config, Box<dyn Error>> {
+        let file = File::open(path).expect("Cannot read file");
```
panic 警察です
```suggestion
        let file = File::open(path)?;
```

YukiYuigishi replied:
`expect`で処理するより、`?`で`Err`を返して関数の外で`expect`したほうがいいって感じですか？

kawaemon replied:
せっかく返り値をResultにしてるんだからそうしたほうが良いんじゃないかという

YukiYuigishi replied:
なるほど。ありがとうございます。

kurume-robo committed:
refactor db_controller::add_task()

Collection::insert_manyからCollection::insert_one へ変更

YukiYuigishi committed:
refactor add_task()

mongo::sync::Collectionを引数として取るように変更
この変更により、add_task()実行のたびにコネクションが確立されなくなる(はず)

YukiYuigishi committed:
refactor read_json_config()

impl教へ入信

YukiYuigishi committed:
refactor Task type

Taskにデフォルト値を実装

YukiYuigishi committed:
refactor db_controller

rename db_controller to db

YukiYuigishi commented:
指摘された箇所を修正しました。

nanai10a reviewed (commented):
ええっと, no-commentでresolvedされてると忘れてるかそうでないかの区別がつかないんですけど, どうなんでしょう…？

YukiYuigishi commented:
ご、ごめんなさい......コメントします。

YukiYuigishi committed:
reafactor add_task test and read_json_config()

YukiYuigishi commented:
GitHub Noobなので、使い方を聞きたいんですが、もしかしなくてもresolvedって自分でやるんじゃなくて他の人に押してもらうべきことだったりしますか？

YukiYuigishi committed:
refactor test add_task()

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+    #[derive(Debug, Serialize, Deserialize)]
+    // Task型
+    pub struct Task {
+        #[serde(with = "ts_seconds_option")]
+        pub scheduled_date: Option<chrono::DateTime<Utc>>,
+        #[serde(with = "ts_seconds")]
+        pub register_date: DateTime<Utc>,
```
どちらかに揃えた方が良いのでは, 一応後者( `::register_date` )に合わせてsuggest

```suggestion
        #[serde(with = "ts_seconds_option")]
        pub scheduled_date: Option<DateTime<Utc>>,
        #[serde(with = "ts_seconds")]
        pub register_date: DateTime<Utc>,
```

YukiYuigishi replied:
acc18947d647274631f6860e89d9fa24db41b25d
done

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+            Self {
+                scheduled_date: None,
+                register_date: Utc::now(),
+                title: "".into(),
+                memo: "".into(),
```
気になって調べたので一応共有
`String::new()` と `"".into()` に効率の違いがあるか見た所, 結構ありそう
追えるところまで追ってみました

[`String::new`](https://doc.rust-lang.org/stable/alloc/string/struct.String.html#method.new)
-> [`Vec::new`](https://doc.rust-lang.org/stable/alloc/vec/struct.Vec.html#method.new)

`<String as Into>::<&str>::into`
**=>** [`<&str as From>::<String>::from`](https://doc.rust-lang.org/stable/alloc/string/struct.String.html#impl-From%3C%26%27_+str%3E=)
**=>** [`<str as ToOwned>::<Owned = String>::to_owned`](https://doc.rust-lang.org/stable/alloc/borrow/trait.ToOwned.html)
=> [`str::as_bytes`](https://doc.rust-lang.org/stable/core/primitive.str.html#method.as_bytes)
-> [`core::mem::transmute`](https://doc.rust-lang.org/stable/core/mem/fn.transmute.html)
**=>** [`<[u8] as ToOwned>::<Owned = Vec<u8>>::to_owned`](https://doc.rust-lang.org/stable/alloc/borrow/trait.ToOwned.html)
=> [`[u8]::to_vec`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.to_vec)
-> [`[u8]::to_vec_in`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.to_vec_in)
**=>** [`String::from_utf8_unchecked`](https://doc.rust-lang.org/stable/alloc/string/struct.String.html#method.from_utf8_unchecked)

**TL;DR**: suggestion参照
```suggestion
                title: String::new(),
                memo: String::new(),
```

YukiYuigishi replied:
TL;DR助かる
acc18947d647274631f6860e89d9fa24db41b25d
done

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+            memo: "Life Love Peace".into(),
+            ..Default::default()
+        };
+        let client = Client::with_uri_str("mongodb://localhost:27017").expect("Cannot Connect DB");
```
```suggestion
        let client = Client::with_uri_str("mongodb://localhost:27017").expect("Cannot connect to DB");
```

YukiYuigishi replied:
修正
df4af8b6cd9c4489a845f0363577a2aadd67d7f1

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+            memo: "Life Love Peace".into(),
+            ..Default::default()
+        };
+        let client = Client::with_uri_str("mongodb://localhost:27017").expect("Cannot Connect DB");
```
```suggestion
        let client = Client::with_uri_str("mongodb://localhost:27017").expect("Cannot connect to DB");
```

YukiYuigishi replied:
修正
df4af8b6cd9c4489a845f0363577a2aadd67d7f1

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+    #[test]
+    fn read_config_from_json() {
+        let config: Config =
+            read_json_config("test/config.json").expect("Cannot read Json config.");
```
（いやUPPERかlowerやろ！などと思いましたがそれ以上に *json config* は違うっぽかったのでsuggestion

```suggestion
            read_json_config("test/config.json").expect("Cannot read `test/config.json`.");
```

YukiYuigishi replied:
修正
 02efb3f14b8f0e4706c71c3b0f12986b70c9fb86

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+        config::{read_json_config, Config},
+        db::{add_task, Task},
+    };
+    use chrono::{TimeZone, Utc};
```
```suggestion
    use chrono::Utc;
```

YukiYuigishi replied:
>Makes a new Date from year, month, day and the current time zone. This assumes the proleptic Gregorian calendar, with the year 0 being 1 BCE.
>The time zone normally does not affect the date (unless it is between UTC-24 and UTC+24), but it will propagate to the DateTime values constructed via this date.
>
>Panics on the out-of-range date, invalid month and/or day.

>[Example](https://docs.rs/chrono/0.4.19/chrono/offset/trait.TimeZone.html#example)
```rs
use chrono::{UTC, TimeZone};
```

らしいので、TimeZoneが必須って認識でいいんですかね？

nanai10a replied:
あ〜〜〜 `TimeZone` ってtraitか, これは失礼しました
Exampleへのリンクくそ古いのだけ気になりましたが… ちゃんとversionに合ったdocを読んでいますか？ (このPjで使われているchronoは [**`0.4.19`**](https://docs.rs/chrono/0.4.19/chrono/offset/trait.TimeZone.html#example) です…)

YukiYuigishi replied:
docをどこで読めばいいかを実はちゃんと理解してなかった。
[crates.io](https://crates.io)じゃなくて[docs](https://docs.rs)で読めばいいって認識であってますか？

nanai10a replied:
https://docs.rs がdocsの置かれてる場所なので, 何も考えずにアドレスバーに `docs.rs/<crate_name>` と打ち込みましょう latestのdocへのaliasになってます
ちなみに標準ライブラリにも繋がってます, ほら: https://docs.rs/std

YukiYuigishi replied:
ありがとうございます。

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+        fn default() -> Self {
+            Self {
+                scheduled_date: None,
+                register_date: Utc::now(),
```
ほんまにそれ [`Utc::new`](https://docs.rs/chrono/latest/chrono/offset/struct.Utc.html#method.now) でええんか？ タイムゾーン的なものは…？？？
と思いましたが杞憂を含むのでどちらでも 代替案は [`Local::new`](https://docs.rs/chrono/latest/chrono/offset/struct.Local.html#method.now)

YukiYuigishi replied:
utcで記録しといて、表示するときにTimeZoneになるようにオフセットかけようかなって考えてます

nanai10a replied:
成程, 了解です.

nanai10a commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+}
+
+pub mod db {
+    use chrono::{serde::ts_seconds, serde::ts_seconds_option, DateTime, Utc};
```
入れ子にuseするならこうじゃないすかね

```suggestion
    use chrono::{serde::{ts_seconds, ts_seconds_option}, DateTime, Utc};
```

YukiYuigishi replied:
acc18947d647274631f6860e89d9fa24db41b25d
done

nanai10a reviewed (changes requested):
あの…なんかパッと見たら色々気になってきたので投げます
あの, ~~正直遊び半分なので気にしなくて結構です~~ ともかく見るだけ見ていただけると

YukiYuigishi committed:
reafactor: test add_new_task_schedule_none

YukiYuigishi committed:
refactor: test add_new_task_scheduled()

YukiYuigishi committed:
refactor: tests typo

YukiYuigishi committed:
refactor: test read_config_from_json error message

kurume-robo committed:
Merge remote-tracking branch 'origin/feat/backend' into feat/backend

kurume-robo committed:
refactor lib.rs

nanai10a reviewed (approved):
確認した限り言う事は無いと思います. お疲れ様でした. ✅

mirror-kt commented on src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+    use mongodb::sync::Client;
+
+    #[test]
+    fn add_new_task_schedule_none() {
```
事後条件をきちんとテストしてください。
`add_task`の呼び出しに成功することは保証していても、呼び出し後にタスクが正しくスケジュールされていることを保証できていません。

mirror-kt commented on src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    #[test]
+    fn add_new_task_scheduled() {
```
事後条件をきちんとテストしてください。
add_taskの呼び出しに成功することは保証していても、呼び出し後にタスクが正しくスケジュールされていることを保証できていません。

MikuroXina commented on src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    //新規タスクの追加
+    pub fn add_task(
+        new_task: Task,
+        collection: &Collection<Task>,
+    ) -> Result<(), mongodb::error::Error> {
+        collection.insert_one(new_task, None)?;
+        Ok(())
+    }
```
これは非常に MongoDB に特化した処理であり, この後にタスクの削除などが追加されることも考えられます. そのため, 以下のようにすると今後の変更がスムーズになるでしょう.

1. `Collection` を保持した `struct` を用意する
2. アプリケーションとして必要なデータベースへの操作 (現状は新規タスクの追加のみ) を表現した `trait` を用意する
3. 1 の `struct` に対して 2 の `trait` を実装する

YukiYuigishi committed:
chore: add thiserror Crate

YukiYuigishi committed:
chore(lib.rs): Separated Config type into a separate file

YukiYuigishi committed:
chore(DB)!: Mongodb was detached from the domain layer

- Task type was separated from lib.rs
- MonoglossicRepository is depend on MonoglossicMongoRepository.
